pkg/controller/pubsub/pubsub_config_controller.go

```diff
@@ -137,9 +138,18 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 	if err != nil {
 		return reconcile.Result{}, err
 	}
-	err = r.system.UpsertConnection(ctx, config, request.Name, cfg.Data["provider"])
-	if err != nil {
-		return reconcile.Result{}, err
+	if newConnFunc, ok := prvd.List()[cfg.Data["provider"]]; ok {
+		err = r.system.UpsertConnection(ctx, config, request.Name, cfg.Data["provider"], newConnFunc)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	} else {
+		err := r.system.CloseConnection(request.Name)
```

Comment:
Why do we need to close the connection on error? What are the failure behaviors if:

1) This is a completely new connection.

2) This is an attempt to change an existing connection.

?

Reply:
This connection is only getting close when there is an attempt to update the existing connection with an unsupported provider. This makes sense to be consistent with connection config and overall behavior. 

If this is a completely new connection requesting to use a provider that is not supported, than we do not initiate any connections and line 147 is no-op.

If this is an attempt to change an existing connection with a provider that is not supported, then we close the connection which needs to be updated to be consistent with it's connection config, as we cannot open connections for other providers.

Does this make sense?

Reply:
> This makes sense to be consistent with connection config and overall behavior.

Is it better to completely remove a connection if a user updates a config to an unsupported flavor, or maintain the old, functional operation (reporting an error on the resource)?

IIRC constraints/templates default to persisting the original state in order to attempt to maintain enforcement.

> If this is a completely new connection requesting to use a provider that is not supported, than we do not initiate any connections and line 147 is no-op.

Agree that net new connection is the trivial case.

> If this is an attempt to change an existing connection with a provider that is not supported, then we close the connection which needs to be updated to be consistent with it's connection config, as we cannot open connections for other providers.

Another possibility is mentioned above.

The current approach more faithfully represents the question: "what would happen if a new pod were to read this config from scratch", the other tries to avoid immediate disruption of messages on failure (at some risk of users being negatively surprised if a pod restarts).

A second argument for not immediately removing-on-error: if we are merely re-reconciling a pre-existing config (i.e. there are no actual config changes), and Upsert() is flaky for whatever reason (maybe it depends on network connnectivity), then we are removing an otherwise valid connection, which may be surprising.

I'd tend to lean toward practicality and robustness-in-the-face-of-flakiness, as opposed to always exactly matching intent.

Curious to hear others' thoughts.

Reply:
> A second argument for not immediately removing-on-error: if we are merely re-reconciling a pre-existing config (i.e. there are no actual config changes), and Upsert() is flaky for whatever reason (maybe it depends on network connnectivity), then we are removing an otherwise valid connection, which may be surprising.

If this is a flaky reconciliation and there are no changes in config then the code to close the connection is not getting executed, as we are closing the connection right now only when one of the following is true:
- config map for an open connection is deleted
- updated config-map for the valid connection has unsupported provider
- updated config-map for the valid connection has supported provider, then we close the old valid connection and open a new valid connection using another provider. For example, I have an audit connection open that uses dapr, but I modify that connection to use rabbitmq, then we close the connection with dapr and open one with rabbitmq for the same config.

When it comes to a flaky reconciliation, right now the same connection is getting updated in place. We could put a check in to determine if this is a flaky reconcile and choose not to update any connections, if that is a good idea to deal with this.

Reply:
As-written, the code calls UpsertConnection() regardless of whether the config has changed or not.

Are you saying that for this code:

```golang
   err1 := r.system.UpsertConnection(ctx, config, request.Name, cfg.Data["provider"], newConnFunc)
   
   err2 := r.system.UpsertConnection(ctx, config, request.Name, cfg.Data["provider"], newConnFunc)
```

`err2` will always be `nil` assuming `err1` is `nil`?

If not, then this above statement is not necessarily true:

> right now the same connection is getting updated in place.

As to this statement:

> We could put a check in to determine if this is a flaky reconcile and choose not to update any connections

We can't necessarily detect if there is a flaky reconcile (we have no historical call data, and even if we did, the harm occurs on the first delete). We can detect if the original state has already been reconciled by caching the data. It seems cheaper and easier to not wipe-on-failure.

Reply:
> err2 will always be nil assuming err1 is nil?

Yip! Unless of course there happens to be a transient error on the driver's side while upserting the connection 2nd time. 

The transient error case might differ from driver to driver. For example, when updating an existing Dapr connection, it typically involves updating the pubsub component used for publishing messages on the already open client, rather than closing the current client and opening a new one. Therefore, for dapr, there won't be any transient errors while upserting the connection. 

However, other drivers may require closing the current client and opening a new one when updating the connection where there is a possibility of getting a transient error. In such cases, it is possible to design the "UpdateConnection" function in the driver to be idempotent. This means that if a dummy update is performed, it will not have any impact on the system or the currently open connection.

Reply:
> We can't necessarily detect if there is a flaky reconcile (we have no historical call data, and even if we did, the harm occurs on the first delete). We can detect if the original state has already been reconciled by caching the data. It seems cheaper and easier to not wipe-on-failure.

Driver have the data related to the currently open connection in each driver's struct (we unmarshal data into the driver struct while initiating the connection), so whenever `UpdateConnection` is invoked, driver should be able to figure out if there is a diff between currently open client config and update request config, and then make the decision accordingly

Reply:
> Yip! Unless of course there happens to be a transient error on the driver's side while upserting the connection 2nd time.

The "unless" clause you give was my point. In general, it's better to program defensively. At scale, something with a 1% likelihood of happening in an individual instance has a >60% chance of happening at least once after 100 instances (1 - 0.99 ** 100).

> The transient error case might differ from driver to driver.

Unless we put a design requirement into the contract of how drivers should be implemented ("do not fail when instantiating a new connection"), we have to allow for the possibility that drivers will fail quasi-randomly. Since we are potentially dealing with network connections, it would be unreasonable to require "never fail".

Put another way, if the case can vary, write for the lowest variation, not the highest.

> Therefore, for dapr, there won't be any transient errors while upserting the connection.

We should not be designing only for dapr. Otherwise, there would be no need for the extra interface work. We would use the dapr client directly.

> However, other drivers may require closing the current client and opening a new one when updating the connection where there is a possibility of getting a transient error. In such cases, it is possible to design the "UpdateConnection" function in the driver to be idempotent. This means that if a dummy update is performed, it will not have any impact on the system or the currently open connection.

This seems like a good idea for the driver author, if for no other reason than to avoid unnecessary handshake traffic. However, we should not assume this will always be possible (or successfully implemented). Allowing for the possibility of failure on our end simplifies our contract with users and hardens our code against unexpected bugs (both ours and driver authors').

> Driver has the data related to the currently open connection in each driver's struct (we unmarshal data into the driver struct while initiating the connection), so whenever UpdateConnection is invoked, driver should be able to figure out if there is a diff between currently open client config and update request config, and then make the decision accordingly

That is what I meant when I said " We can detect if the original state has already been reconciled by caching the data." being pedantic, a re-sync of already synced data is not a flaky reconcile. A "flake" is a repeated operation that has unpredictable results, with some of those results being undesireable (e.g. a "flaky test").

Whether we should expect driver authors to handle this use case correctly (and whether it's wise to assume they can always do so effectively) is addressed above.

Reply:
Ahh, this makes sense! So the idea is to have proofing on the system side to ensure we do not close any active connection (when we are replacing/updating) unless we can open a connection with the updated config. This leaves us less prune to flakiness and transient errors than wipe-on-failure. Additionally, I should probably change the order of execution to initiate the connection -> close old connection if exists -> update connection map in `UpsertConnection`, to make sure we have an active connection before closing an old one.

Reply:
Nice!